File: python/Caesar.py
# -*- coding: utf-8 -*-
#
# Caesar cipher
# 
# @author  lellansin <lellansin@gmail.com>
# @website http://www.lellansin.com/tutorials/ciphers
#

a, z = ord('a'), ord('z')
A, Z = ord('A'), ord('Z')

# 
# 凯撒密码
# 
def caesar(words, shift):
    cipher = ''

    for ch in words:
        if ch.islower():
            cipher += getShiftCh(ch, a, z, shift)
        elif ch.isupper():
            cipher += getShiftCh(ch, A, Z, shift)
        else:
            cipher += ch
    return cipher


def getShiftCh(ch, A, Z, shift):
    result = dif = ord(ch) + shift
    if dif > Z:
        result = (dif - Z) % 26 - 1 + A
    elif dif < A:
        result = Z - ((A - dif) % 26 - 1)
    return chr(result)


if __name__ == '__main__':
    text = 'hello world, this is Caesar cipher.'
    
    # 右移
    ciphertext = caesar(text, 3);
    print(ciphertext)

    # 左移
    print(caesar(ciphertext, -3))

File: javascript/Caesar.js
/*
 * Caesar Cipher
 *
 * @author  lellansin <lellansin@gmail.com>
 * @website http://www.lellansin.com/tutorials/ciphers
 */
(function() {

    var ASCII = {
        a: 'a'.charCodeAt(),
        z: 'z'.charCodeAt(),
    };

    function caesar(words, shift) {
        return words.toLowerCase().replace(/\w/g, function(ch) {
            return getShiftCh(ch, shift);
        });
    };

    var getShiftCh = function(ch, shift) {
        var asc = ch.charCodeAt() + shift;
        var new_ch = asc;
        if (asc > ASCII.z) {
            new_ch = ((asc - ASCII.a) % 26) + ASCII.a;
        } else if (asc < ASCII.a) {
            new_ch = ASCII.z - ((ASCII.a - asc) % 26 - 1);
        }
        return String.fromCharCode(new_ch);
    };

    /* -------------------- 测试 -------------------- */

    var text = 'hello world, this is Caesar cipher.';
    var shift = 3;

    var ciphertext = caesar(text, shift);
    console.log('右移 [%s]', ciphertext);
    console.log('左移 [%s]', caesar(ciphertext, -shift));

}());
